lib/conflicts/conflicts/core/pipeline.py

```diff
@@ -226,25 +236,39 @@ def process_document_pair(self, document_pair: DocumentPair) -> Tuple[bool, Dict
                 self.logger.warning("Validation failed, retrying...")
                 time.sleep(1)
 
-        # Step 3: Save to database if validation passed
-        if validation_result and validation_result.is_valid:
+        # Step 4: Save all attempts to database (both successful and failed)
+        saved_attempts = []
+        for editor_result_attempt, validation_result_attempt, attempt_num in all_attempts:
             is_success = self._save_to_database(
-                pair_id,
+                f"{pair_id}_attempt_{attempt_num}",
                 document_pair,
-                editor_result,
+                editor_result_attempt,
                 conflict_result.conflict_type,
-                validation_result,
+                validation_result_attempt,
+            )
+            saved_attempts.append(
+                {
+                    "attempt": attempt_num,
+                    "saved": is_success,
+                    "valid": validation_result_attempt.is_valid,
+                    "score": validation_result_attempt.score,
+                }
             )
-            result_data["success"] = is_success
+
+        result_data["all_attempts"] = saved_attempts
+        result_data["success"] = validation_result.is_valid if validation_result else False
 
         result_data["processing_time"] = time.time() - start_time
 
         # Summary log
         status = "SUCCESS" if result_data["success"] else "FAILED"
+        total_attempts = len(saved_attempts)
+        successful_attempts = sum(1 for attempt in saved_attempts if attempt["valid"])
+
         self.logger.info(
             f"Pair {pair_id}: {status} - {conflict_result.conflict_type} conflict, "
             f"{proposition_result[0].total_propositions + proposition_result[1].total_propositions}"
-            f" propositions"
+            f" propositions, {successful_attempts}/{total_attempts} attempts saved"
```

Comment:
to my mind, you can merge the step 3 and 4 in the same loop it simplifies the code